File: examples/python/minimal/agent.py
"""
Minimal Lilith reference implementation — static allow/deny.

Demonstrates:
  1. SDK initialisation (binary auto-discovery via LILITH_ZERO_BINARY_PATH or PATH)
  2. Tool discovery (list_tools)
  3. Calling an allowed tool
  4. Calling a denied tool → PolicyViolationError
  5. Session ID — HMAC-signed identifier accessible as lilith.session_id

Run:
    export LILITH_ZERO_BINARY_PATH=/path/to/lilith-zero   # or add to PATH
    python agent.py
"""

import asyncio
import os
import sys

from lilith_zero import Lilith, PolicyViolationError

POLICY = os.path.join(os.path.dirname(__file__), "policy.yaml")
SERVER = os.path.join(os.path.dirname(__file__), "server.py")


async def main() -> None:
    # Lilith auto-discovers its binary via LILITH_ZERO_BINARY_PATH env var,
    # PATH, or the standard user install location (~/.lilith_zero/bin).
    async with Lilith(
        upstream=f"{sys.executable} -u {SERVER}",
        policy=POLICY,
    ) as lilith:
        print(f"session  : {lilith.session_id}")

        # --- Tool discovery ---
        tools = await lilith.list_tools()
        print(f"tools    : {[t['name'] for t in tools]}")

        # --- Allowed calls ---
        result = await lilith.call_tool("search_web", {"query": "Lilith Zero MCP security"})
        print(f"search   : {result['content'][0]['text']}")

        result = await lilith.call_tool("get_time", {})
        print(f"time     : {result['content'][0]['text']}")

        # --- Denied call ---
        try:
            await lilith.call_tool("query_database", {"sql": "SELECT * FROM users"})
            print("ERROR: should have been blocked")
        except PolicyViolationError as exc:
            print(f"blocked  : {exc}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/ts/minimal/agent.ts
/**
 * TypeScript minimal Lilith example — stdio transport.
 *
 * Mirrors examples/python/minimal/agent.py exactly:
 *   1. SDK init (binary auto-discovery)
 *   2. Tool discovery
 *   3. Allowed call
 *   4. Denied call → PolicyViolationError
 *   5. Session ID + audit log
 *
 * Run:
 *   export LILITH_ZERO_BINARY_PATH=/path/to/lilith-zero
 *   bun agent.ts
 */

import { join } from "path";
import { Lilith, PolicyViolationError } from "../../../sdk-ts/src/index.ts";

const POLICY = join(import.meta.dir, "policy.yaml");
const SERVER = join(import.meta.dir, "server.mjs");

const lilith = new Lilith({
  upstream: `node ${SERVER}`,
  policy: POLICY,
});

await using _ = lilith;
await lilith.connect();

console.log(`session  : ${lilith.sessionId}`);

const tools = await lilith.listTools();
console.log(`tools    : [${tools.map((t) => t.name).join(", ")}]`);

const search = await lilith.callTool("search_web", { query: "Lilith Zero MCP security" });
console.log(`search   : ${search.content[0].text}`);

const time = await lilith.callTool("get_time", {});
console.log(`time     : ${time.content[0].text}`);

try {
  await lilith.callTool("query_database", { sql: "SELECT * FROM users" });
  console.log("ERROR: should have been blocked");
} catch (e) {
  if (e instanceof PolicyViolationError) {
    console.log(`blocked  : ${e.message}`);
  } else throw e;
}

const audit = await lilith.drainAuditLogs();
console.log(`\naudit    : ${audit.length} entries`);
for (const entry of audit) {
  const dec = entry.details?.decision ?? entry.event_type;
  const tool = entry.details?.tool_name ?? "";
  console.log(`  [${dec?.padEnd(6)}] ${tool}`);
}
